app/src/main/java/org/mozilla/focus/locale/screen/LanguageMiddleware.kt

```diff
@@ -0,0 +1,65 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.locale.screen
+
+import android.app.Activity
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.locale.LocaleManager
+import mozilla.components.support.locale.LocaleUseCases
+import org.mozilla.focus.locale.Locales
+import org.mozilla.focus.settings.InstalledSearchEnginesSettingsFragment
+import org.mozilla.gecko.util.ThreadUtils.runOnUiThread
+import java.util.Locale
+
+class LanguageMiddleware(val activity: Activity, private val localeUseCase: LocaleUseCases) :
+    Middleware<LanguageScreenState, LanguageScreenAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<LanguageScreenState, LanguageScreenAction>,
+        next: (LanguageScreenAction) -> Unit,
+        action: LanguageScreenAction
+    ) {
+        when (action) {
+            is LanguageScreenAction.Select -> {
+                LanguageStorage(context = activity).saveCurrentLanguageInSharePref(action.selectedLanguage.tag)
+                setCurrentLanguage(action.selectedLanguage.tag)
+            }
+            is LanguageScreenAction.UpdateLanguages -> {
+                next(action)
+            }
+            is LanguageScreenAction.InitLanguages -> {
+                /**
+                 * The initial LanguageScreenState when the user enters first in the screen
+                 */
+                val storage = LanguageStorage(context = activity)
+                context.dispatch(
+                    LanguageScreenAction.UpdateLanguages(
+                        storage.getLanguages(),
+                        storage.getSelectedLanguageTag()
+                    )
+                )
+            }
```

Comment:
In the `else` case we may want to invoke `next()`, otherwise this middleware will intercept all other (future) actions.